File: golang/2612_min_reverse_operations/min_reverse_operations.go
// Package main ...
package main

import (
	"slices"
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
)

func minReverseOperations(n int, p int, banned []int, k int) []int {
	ban := make(map[int]bool)
	for _, b := range banned {
		ban[b] = true
	}
	sets := make([][]int, 2)
	for i := 0; i < n; i++ {
		if i != p && !ban[i] {
			sets[i%2] = append(sets[i%2], i)
		}
	}
	for i := range sets {
		sort.Ints(sets[i])
	}
	ans := make([]int, n)
	for i := range ans {
		ans[i] = -1
	}
	q := []int{p}
	ans[p] = 0
	for len(q) > 0 {
		i := q[0]
		q = q[1:]
		mn := max(i-k+1, k-i-1)
		mx := min(i+k-1, 2*n-k-i-1)
		targetSet := sets[mx%2]
		toRemove := []int{}
		left := sort.SearchInts(targetSet, mn)
		right := sort.SearchInts(targetSet, mx+1)
		for j := left; j < right; j++ {
			val := targetSet[j]
			ans[val] = ans[i] + 1
			q = append(q, val)
			toRemove = append(toRemove, val)
		}
		for _, val := range toRemove {
			idx := sort.SearchInts(targetSet, val)
			if idx < len(targetSet) && targetSet[idx] == val {
				targetSet = slices.Delete(targetSet, idx, idx+1)
			}
		}
		sets[mx%2] = targetSet
	}

	return ans
}

func main() {
	tests := []struct {
		n      int
		p      int
		banned []int
		k      int
		ans    []int
	}{
		{4, 0, []int{1, 2}, 4, []int{0, -1, -1, 1}},
		{5, 0, []int{2, 4}, 3, []int{0, -1, -1, -1, -1}},
		{4, 2, []int{0, 1, 3}, 1, []int{-1, -1, 0, -1}},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, minReverseOperations(test.n, test.p, test.banned, test.k), index)
	}
}

File: rust/2612_min_reverse_operations/min_reverse_operations.rs
struct Solution;

impl Solution {
    pub fn min_reverse_operations(n: i32, p: i32, banned: Vec<i32>, k: i32) -> Vec<i32> {
        use std::collections::{BTreeSet, HashSet, VecDeque};
        let n = n as usize;
        let p = p as usize;
        let k = k as usize;
        let ban: HashSet<i32> = banned.into_iter().collect();
        let mut sets = vec![BTreeSet::new(), BTreeSet::new()];
        for i in 0..n {
            if i != p && !ban.contains(&(i as i32)) {
                sets[i % 2].insert(i as i32);
            }
        }
        let mut ans = vec![-1; n];
        let mut q = VecDeque::new();
        q.push_back(p as i32);
        ans[p] = 0;
        while let Some(i) = q.pop_front() {
            let i = i as usize;
            let mn = (i as i32 - k as i32 + 1).max(k as i32 - i as i32 - 1) as usize;
            let mx = (i as i32 + k as i32 - 1).min(n as i32 * 2 - k as i32 - i as i32 - 1) as usize;
            let set = &mut sets[mx % 2];
            let mut to_remove = Vec::new();
            for &val in set.range(mn as i32..=mx as i32) {
                ans[val as usize] = ans[i] + 1;
                q.push_back(val);
                to_remove.push(val);
            }
            for val in to_remove {
                set.remove(&val);
            }
        }
        ans
    }
}

fn main() {
    let tests = vec![
        (4, 0, vec![1, 2], 4, vec![0, -1, -1, 1]),
        (5, 0, vec![2, 4], 3, vec![0, -1, -1, -1, -1]),
        (4, 2, vec![0, 1, 3], 1, vec![-1, -1, 0, -1]),
    ];

    for (n, p, banned, k, ans) in tests {
        assert_eq!(Solution::min_reverse_operations(n, p, banned, k), ans);
    }
}
